In [1]:
import numpy as np
import pandas as pd

Let's start with the batters file. We'll read that first.

Also, here's a list of potential players I want to investigate further:'
Position Players:
- Matt Kemp
- Mike Trout
- Jetes
- Giancarlo?
- Jose Altuve
- Ichiro
- Barry Bonds

Pitchers:
- Kersh
- Thor
- Greinke
- Kuroda
- Arrieta


In [2]:
batters = pd.read_table("batters_cleaned.tsv")

In [22]:
batters.shape

(3625, 22)

In [78]:
batters.describe()

year              name      position  team games    pa    ab  runs  \
count   3500              3500          3500  3500  3467  3500  3500  3231   
unique   140               163            54   204   179   715   652   167   
top     1928  Rickey Henderson  Rightfielder   PIT   152   633   564     0   
freq      53                41           317   215    94    21    21    64   

        hits    hr   rbi    sb    cs    bb    so     ba    obp   slg   ops  \
count   3500  3500  3196  3141  3500  3200  3155   3500   3500  3500  3500   
unique   252    58   172    97    36   146   148    281    325   513   706   
top        1     0     0     0     0     0    17  0.309  0.352     0     0   
freq      39   490    75   476  1528    82    65     45     42    35    29   

          tb   ibb  
count   3500  3500  
unique   407    35  
top        0     0  
freq      35  2303

In [3]:
batters.head()

year          name       position     team games   pa   ab runs hits  hr  \
0  1990  Jeff Bagwell  First Baseman  BOS-min   136  569  481   63  160   4   
1  1991  Jeff Bagwell  First Baseman      HOU   156  650  554   79  163  15   
2  1992  Jeff Bagwell  First Baseman      HOU   162  697  586   87  160  18   
3  1993  Jeff Bagwell  First Baseman      HOU   142  609  535   76  171  20   
4  1994  Jeff Bagwell  First Baseman      HOU   110  479  400  104  147  39   

  ... cs  bb   so     ba    obp    slg    ops ops_plus   tb ibb  
0 ...  7  73   57  0.333  0.422  0.457   0.88     None  220  12  
1 ...  4  75  116  0.294  0.387  0.437  0.824      139  242   5  
2 ...  6  84   97  0.273  0.368  0.444  0.812      135  260  13  
3 ...  4  62   73   0.32  0.388  0.516  0.903      145  276   6  
4 ...  4  65   65  0.368  0.451   0.75  1.201      213  300  14  

[5 rows x 22 columns]

Let's replace any of the "None" values with na's, that way we can get an accurate number of missing values.

In [13]:
batters = batters.replace('None', np.nan)

Now we want to figure out where the missing data is. We can use the isnull method.

In [14]:
batters.isnull()

year   name position   team  games     pa     ab   runs   hits     hr  \
0     False  False    False  False  False  False  False  False  False  False   
1     False  False    False  False  False  False  False  False  False  False   
2     False  False    False  False  False  False  False  False  False  False   
3     False  False    False  False  False  False  False  False  False  False   
4     False  False    False  False  False  False  False  False  False  False   
5     False  False    False  False  False  False  False  False  False  False   
6     False  False    False  False  False  False  False  False  False  False   
7     False  False    False  False  False  False  False  False  False  False   
8     False  False    False  False  False  False  False  False  False  False   
9     False  False    False  False  False  False  False  False  False  False   
10    False  False    False  False  False  False  False  False  False  False   
11    False  False    False  False  False  False  False  False  False  False   
12    False  False    False  False  False  False  False  False  False  False   
13    False  False    False  False  False  False  False  False  False  False   
14    False  False    False  False  False  False  False  False  False  False   
15    False  False    False  False  False  False  False  False  False  False   
16    False  False    False  False  False  False  False  False  False  False   
17    False  False    False  False  False  False  False  False  False  False   
18    False  False    False  False   True  False  False  False  False  False   
19    False  False    False  False   True  False  False  False  False  False   
20    False  False    False  False   True  False  False  False  False  False   
21    False  False    False  False   True  False  False  False  False  False   
22    False  False    False  False   True  False  False  False  False  False   
23    False  False    False  False   True  False  False  False  False  False   
24    False  False    False  False   True  False  False  False  False  False   
25    False  False    False  False   True  False  False  False  False  False   
26    False  False    False  False   True  False  False  False  False  False   
27    False  False    False  False   True  False  False  False  False  False   
28    False  False    False  False   True  False  False  False  False  False   
29    False  False    False  False  False  False  False  False  False  False   
...     ...    ...      ...    ...    ...    ...    ...    ...    ...    ...   
3595  False  False    False  False  False  False  False  False  False  False   
3596  False  False    False  False  False  False  False  False  False  False   
3597  False  False    False  False  False  False  False  False  False  False   
3598  False  False    False  False  False  False  False  False  False  False   
3599  False  False    False  False  False  False  False  False  False  False   
3600  False  False    False  False  False  False  False  False  False  False   
3601  False  False    False  False  False  False  False  False  False  False   
3602  False  False    False  False  False  False  False  False  False  False   
3603  False  False    False  False  False  False  False  False  False  False   
3604  False  False    False  False  False  False  False  False  False  False   
3605  False  False    False  False  False  False  False  False  False  False   
3606  False  False    False  False  False  False  False  False  False  False   
3607  False  False    False  False  False  False  False  False  False  False   
3608  False  False    False  False  False  False  False  False  False  False   
3609  False  False    False  False  False  False  False  False  False  False   
3610  False  False    False  False  False  False  False  False  False  False   
3611  False  False    False  False  False  False  False  False  False  False   
3612  False  False    False  False  False  False  False  False  False  False   
3613  F

Get column names (out of curiousity)

In [15]:
list(batters)

['year',
 'name',
 'position',
 'team',
 'games',
 'pa',
 'ab',
 'runs',
 'hits',
 'hr',
 'rbi',
 'sb',
 'cs',
 'bb',
 'so',
 'ba',
 'obp',
 'slg',
 'ops',
 'ops_plus',
 'tb',
 'ibb']

Sum up the boolean array to see how many missing values each column has. Remember shape is (3625, 22)

In [101]:
np.sum(batters.isnull())

year          0
name          0
position      0
team          0
games        33
pa            0
ab            0
runs        269
hits          0
hr            0
rbi         304
sb          359
cs            0
bb            0
so            0
ba            0
obp           0
slg           0
ops           0
tb            0
ibb           0
bb_rate       0
so_rate       0
dtype: int64

In [23]:
batters.loc[batters.isnull().any(axis=1),:]

year              name                                  position  \
0     1990      Jeff Bagwell                             First Baseman   
5     1995      Jeff Bagwell                             First Baseman   
16    2005      Jeff Bagwell                             First Baseman   
18    1936     Willard Brown                              Rightfielder   
19    1937     Willard Brown                              Rightfielder   
20    1938     Willard Brown                              Rightfielder   
21    1939     Willard Brown                              Rightfielder   
22    1940     Willard Brown                              Rightfielder   
23    1941     Willard Brown                              Rightfielder   
24    1942     Willard Brown                              Rightfielder   
25    1943     Willard Brown                              Rightfielder   
26    1944     Willard Brown                              Rightfielder   
27    1946     Willard Brown                              Rightfielder   
28    1947     Willard Brown                              Rightfielder   
29    1947     Willard Brown                              Rightfielder   
30    1948     Willard Brown                              Rightfielder   
31    1949     Willard Brown                              Rightfielder   
32    1950     Willard Brown                              Rightfielder   
33    1951     Willard Brown                              Rightfielder   
34    1953     Willard Brown                              Rightfielder   
35    1954     Willard Brown                              Rightfielder   
36    1955     Willard Brown                              Rightfielder   
37    1956     Willard Brown                              Rightfielder   
38    1979        Cal Ripken               Shortstop and Third Baseman   
39    1980        Cal Ripken               Shortstop and Third Baseman   
40    1981        Cal Ripken               Shortstop and Third Baseman   
62    1982        Tony Gwynn                              Rightfielder   
64    1983        Tony Gwynn                              Rightfielder   
84    1913  Billy Southworth                              Rightfielder   
85    1913  Billy Southworth                              Rightfielder   
...    ...               ...                                       ...   
3444  1891      Deacon White   Third Baseman, Catcher and Rightfielder   
3446  1961         Joe Torre  Catcher, First Baseman and Third Baseman   
3464  1990      Frank Thomas       Designated Hitter and First Baseman   
3480  2005      Frank Thomas       Designated Hitter and First Baseman   
3487  1988      Craig Biggio    Second Baseman, Catcher and Outfielder   
3508  1990       Mike Piazza                 Catcher and First Baseman   
3509  1991       Mike Piazza                 Catcher and First Baseman   
3510  1992       Mike Piazza                 Catcher and First Baseman   
3525  2003       Mike Piazza                 Catcher and First Baseman   
3527  2004       Mike Piazza                 Catcher and First Baseman   
3531  2007       Mike Piazza                 Catcher and First Baseman   
3533  1988       Ken Griffey                             Centerfielder   
3540  1995       Ken Griffey                             Centerfielder   
3559  1990      Ivn Rodrguez                                   Catcher   
3560  1991      Ivn Rodrguez                                   Catcher   
3572  2002      Ivn Rodrguez                                   Catcher   
3579  2008      Ivn Rodrguez                                   Catcher   
3583  2009      Ivn Rodrguez                                   Catcher   
3587  2010      Ivn Rodrguez                                   Catcher   
3589  2011      Ivn Rodrguez                                   Catcher   
3591  1978        Tim Raines                               Leftfielder   
3592  1979        Tim Raines                               Leftfielder   
3593  1979

Let's make a list. The most missing values is IBB with 2427 out of 3625. It could be that at the time, IBBs weren't being recorded. Or, if the player actually wasn't intentionally, it was left blank. I'm going to replace these with zero.

In [30]:
batters['ibb'].fillna(0)

0       12
1        5
2       13
3        6
4       14
5        1
6       12
7       20
8       27
9        8
10      16
11      11
12       5
13       8
14       3
15       6
16       0
17       1
18       0
19       0
20       0
21       0
22       0
23       0
24       0
25       0
26       0
27       0
28       0
29       0
        ..
3595     0
3596     5
3597     9
3598     9
3599     7
3600    13
3601     9
3602    26
3603    14
3604    18
3605     8
3606     9
3607     4
3608     0
3609     4
3610     3
3611     3
3612     4
3613     1
3614     2
3615     0
3616     1
3617     1
3618     0
3619     0
3620     0
3621     0
3622     0
3623     0
3624     4
Name: ibb, dtype: object

In [76]:
np.sum(batters.isnull())

year          0
name          0
position      0
team          0
games        33
pa            0
ab            0
runs        269
hits          0
hr            0
rbi         304
sb          359
cs            0
bb          300
so          345
ba            0
obp           0
slg           0
ops           0
tb            0
ibb           0
dtype: int64

Cool. Let's group all columns that I think I can replace missing values with zero. They are cs and tb.

In [36]:
batters['cs'] = batters['cs'].fillna(0)
batters['tb'] = batters['tb'].fillna(0)

I'm curious about some of these missing BA's and OBP's and SLG. Can we impute via data provided?

In [41]:
pd.set_option('display.max_columns', None)
batters.loc[batters.isnull().hits,:]

year              name                                       position  \
33    1951     Willard Brown                                   Rightfielder   
86     NaN  Billy Southworth                                   Rightfielder   
90     NaN  Billy Southworth                                   Rightfielder   
106    NaN  Billy Southworth                                   Rightfielder   
314    NaN     George Sisler                                  First Baseman   
386   1879  Charlie Comiskey                                  First Baseman   
401   1896  Charlie Comiskey                                  First Baseman   
403   1899  Charlie Comiskey                                  First Baseman   
484   1939    Rogers Hornsby    Second Baseman, Shortstop and Third Baseman   
486   1941    Rogers Hornsby    Second Baseman, Shortstop and Third Baseman   
488   1944    Rogers Hornsby    Second Baseman, Shortstop and Third Baseman   
510    NaN      Jim O'Rourke          Outfielder, Catcher and First Baseman   
512   1896      Jim O'Rourke          Outfielder, Catcher and First Baseman   
513   1897      Jim O'Rourke          Outfielder, Catcher and First Baseman   
514   1898      Jim O'Rourke          Outfielder, Catcher and First Baseman   
563   1900   Hughie Jennings                    Shortstop and First Baseman   
569    NaN   Hughie Jennings                    Shortstop and First Baseman   
574    NaN   Hughie Jennings                    Shortstop and First Baseman   
577    NaN   Hughie Jennings                    Shortstop and First Baseman   
579    NaN   Hughie Jennings                    Shortstop and First Baseman   
596    NaN        Hugh Duffy                                     Outfielder   
604   1888      Ed Delahanty  Leftfielder, First Baseman and Second Baseman   
662    NaN       Fred Clarke                                    Leftfielder   
668   1881     Dan Brouthers                                  First Baseman   
686   1896     Dan Brouthers                                  First Baseman   
688    NaN     Dan Brouthers                                  First Baseman   
697    NaN   Roger Bresnahan                      Catcher and Centerfielder   
699   1899   Roger Bresnahan                      Catcher and Centerfielder   
721   1919   Roger Bresnahan                      Catcher and Centerfielder   
722   1920   Roger Bresnahan                      Catcher and Centerfielder   
...    ...               ...                                            ...   
2205  1945       Duke Snider                                  Centerfielder   
2282  1943       Johnny Mize                                  First Baseman   
2283  1944       Johnny Mize                                  First Baseman   
2284  1945       Johnny Mize                                  First Baseman   
2419  1943     Pee Wee Reese                    Shortstop and Third Baseman   
2420  1944     Pee Wee Reese                    Shortstop and Third Baseman   
2421  1945     Pee Wee Reese                    Shortstop and Third Baseman   
2484   NaN      Rick Ferrell                                        Catcher   
2517   NaN      Arky Vaughan                    Shortstop and Third Baseman   
2528  1943    Enos Slaughter                                   Rightfielder   
2529  1944    Enos Slaughter                                   Rightfielder   
2530  1945    Enos Slaughter                                   Rightfielder   
2636  1945       Bobby Doerr                                 Second Baseman   
2828  1968    Reggie Jackson                                   Rightfielder   
2854  1943      Phil Rizzuto                                      Shortstop   
2855  1944      Phil Rizzuto                                      Shortstop   
2856  1945      Phil Rizzuto                                      Shortstop   
2887  1946    Richie Ashburn                                  Centerfielder   
2904  1878        Ned Hanlon                            

At least when checking for why hits are missing, there is no relevant info for why they are missing or why everything is gone. I think it's possibly because Scrapy scraped hidden rows that don't actually show on the webspage, or because these players played in the 1800's and don't have all available data.

In [50]:
batters = batters.dropna(subset=['hits']) 

In [51]:
batters.loc[batters.isnull().any(axis=1),:]

year              name                                  position  \
0     1990      Jeff Bagwell                             First Baseman   
5     1995      Jeff Bagwell                             First Baseman   
16    2005      Jeff Bagwell                             First Baseman   
18    1936     Willard Brown                              Rightfielder   
19    1937     Willard Brown                              Rightfielder   
20    1938     Willard Brown                              Rightfielder   
21    1939     Willard Brown                              Rightfielder   
22    1940     Willard Brown                              Rightfielder   
23    1941     Willard Brown                              Rightfielder   
24    1942     Willard Brown                              Rightfielder   
25    1943     Willard Brown                              Rightfielder   
26    1944     Willard Brown                              Rightfielder   
27    1946     Willard Brown                              Rightfielder   
28    1947     Willard Brown                              Rightfielder   
30    1948     Willard Brown                              Rightfielder   
31    1949     Willard Brown                              Rightfielder   
32    1950     Willard Brown                              Rightfielder   
34    1953     Willard Brown                              Rightfielder   
35    1954     Willard Brown                              Rightfielder   
36    1955     Willard Brown                              Rightfielder   
37    1956     Willard Brown                              Rightfielder   
38    1979        Cal Ripken               Shortstop and Third Baseman   
39    1980        Cal Ripken               Shortstop and Third Baseman   
40    1981        Cal Ripken               Shortstop and Third Baseman   
62    1982        Tony Gwynn                              Rightfielder   
64    1983        Tony Gwynn                              Rightfielder   
84    1913  Billy Southworth                              Rightfielder   
85    1913  Billy Southworth                              Rightfielder   
87    1914  Billy Southworth                              Rightfielder   
88    1915  Billy Southworth                              Rightfielder   
...    ...               ...                                       ...   
3438  1885      Deacon White   Third Baseman, Catcher and Rightfielder   
3444  1891      Deacon White   Third Baseman, Catcher and Rightfielder   
3446  1961         Joe Torre  Catcher, First Baseman and Third Baseman   
3464  1990      Frank Thomas       Designated Hitter and First Baseman   
3480  2005      Frank Thomas       Designated Hitter and First Baseman   
3487  1988      Craig Biggio    Second Baseman, Catcher and Outfielder   
3508  1990       Mike Piazza                 Catcher and First Baseman   
3509  1991       Mike Piazza                 Catcher and First Baseman   
3510  1992       Mike Piazza                 Catcher and First Baseman   
3525  2003       Mike Piazza                 Catcher and First Baseman   
3527  2004       Mike Piazza                 Catcher and First Baseman   
3531  2007       Mike Piazza                 Catcher and First Baseman   
3533  1988       Ken Griffey                             Centerfielder   
3540  1995       Ken Griffey                             Centerfielder   
3559  1990      Ivn Rodrguez                                   Catcher   
3560  1991      Ivn Rodrguez                                   Catcher   
3572  2002      Ivn Rodrguez                                   Catcher   
3579  2008      Ivn Rodrguez                                   Catcher   
3583  2009      Ivn Rodrguez                                   Catcher   
3587  2010      Ivn Rodrguez                                   Catcher   
3589  2011      Ivn Rodrguez                                   Catcher   
3591  1978        Tim Raines                               Leftfielder   
3592  1979

Hmmm I'm not exactly sure what to make of the 33 missing games now, I believe they are MCAR. Leave them as NA, I will primarily be focused more on PA's per season, which we are not missing any of.

In [54]:
batters.loc[batters.isnull().games,:]

year            name                       position team games   pa  \
18    1936   Willard Brown                   Rightfielder  KCM   NaN    1   
19    1937   Willard Brown                   Rightfielder  KCM   NaN  178   
20    1938   Willard Brown                   Rightfielder  KCM   NaN   78   
21    1939   Willard Brown                   Rightfielder  KCM   NaN   72   
22    1940   Willard Brown                   Rightfielder  KCM   NaN    4   
23    1941   Willard Brown                   Rightfielder  KCM   NaN   92   
24    1942   Willard Brown                   Rightfielder  KCM   NaN  133   
25    1943   Willard Brown                   Rightfielder  KCM   NaN   79   
26    1944   Willard Brown                   Rightfielder  KCM   NaN    4   
27    1946   Willard Brown                   Rightfielder  KCM   NaN  109   
28    1947   Willard Brown                   Rightfielder  KCM   NaN   61   
30    1948   Willard Brown                   Rightfielder  KCM   NaN   92   
31    1949   Willard Brown                   Rightfielder  KCM   NaN    3   
1638  1938  Roy Campanella                        Catcher  BEG   NaN   11   
1639  1939  Roy Campanella                        Catcher  BEG   NaN   59   
1640  1940  Roy Campanella                        Catcher  BEG   NaN   87   
1641  1941  Roy Campanella                        Catcher  BEG   NaN   82   
1642  1942  Roy Campanella                        Catcher  BEG   NaN  122   
1643  1943  Roy Campanella                        Catcher  BEG   NaN    4   
1644  1944  Roy Campanella                        Catcher  BEG   NaN   86   
1645  1945  Roy Campanella                        Catcher  BEG   NaN   47   
1872  1939     Monte Irvin  Leftfielder and First Baseman   NE   NaN   57   
1873  1940     Monte Irvin  Leftfielder and First Baseman   NE   NaN  135   
1874  1941     Monte Irvin  Leftfielder and First Baseman   NE   NaN  142   
1875  1942     Monte Irvin  Leftfielder and First Baseman   NE   NaN   12   
1876  1945     Monte Irvin  Leftfielder and First Baseman   NE   NaN   14   
1877  1946     Monte Irvin  Leftfielder and First Baseman   NE   NaN  162   
1878  1947     Monte Irvin  Leftfielder and First Baseman   NE   NaN   53   
1879  1948     Monte Irvin  Leftfielder and First Baseman   NE   NaN   34   
2942  1943      Larry Doby                  Centerfielder   NE   NaN   65   
2943  1944      Larry Doby                  Centerfielder   NE   NaN    5   
2944  1946      Larry Doby                  Centerfielder   NE   NaN  188   
2945  1947      Larry Doby                  Centerfielder   NE   NaN   36   

       ab runs hits hr rbi sb cs  bb   so     ba    obp    slg    ops  \
18      1    0    0  0   0  0  0   0  NaN      0      0      0      0   
19    174   36   58  6  17  6  0   3  NaN  0.333  0.345  0.529  0.873   
20     78   14   27  3   7  9  0   0  NaN  0.346  0.346  0.551  0.897   
21     70   10   22  1   8  5  0   2  NaN  0.314  0.333  0.486  0.819   
22      4    0    0  0   0  0  0   0  NaN      0      0      0      0   
23     87   19   30  2  19  4  0   5  NaN  0.345   0.38  0.517  0.898   
24    127   28   47  5  25  3  0   6  NaN   0.37  0.398  0.575  0.973   
25     79   15   27  3   6  2  0   0  NaN  0.342  0.342  0.494  0.835   
26      4    0    1  0   0  0  0   0  NaN   0.25   0.25   0.25    0.5   
27    106   15   27  2  21  8  0   3  NaN  0.255  0.275  0.443  0.719   
28     57   12   21  2  18  0  0   4  NaN  0.368   0.41  0.702  1.112   
30     86   18   31  3  15  4  0   6  NaN   0.36  0.402  0.547  0.949   
31      3    0    0  0   0  0  0   0  NaN      0      0      0      0   
1638   11    3    4  0   0  0  0   0  NaN  0.364  0.364  0.364  0.727   
1639   55    7   15  1   6  0  0   3  NaN  0.273   0.31  0.345  0.656   
1640   85   12   24  5  21  1  0   2  NaN  0.282  0.299  0.518  0.816   
1641   76   15   24  4  22  1  0   5  NaN  0.316  0.358  0.579  0.937   
1642  114   18   33  1  29  1  0   7  NaN  0.289  0.331  0.395  0.725   


Deleting OPS+ cuz I have no idea what it is and if I can't understand it, how am I going to explain it to people.

In [63]:
batters = batters.drop('ops_plus', 1)

ValueError: labels ['ops_plus'] not contained in axis

There are still 5 columns in which we have over 300 missing values (so, sb, r, rb, bb). Let's look at these

In [66]:
batters.loc[batters.isnull().ba,:]

year              name                                 position team  \
85    1913  Billy Southworth                             Rightfielder  CLE   
576   1910   Hughie Jennings              Shortstop and First Baseman  DET   
580   1918   Hughie Jennings              Shortstop and First Baseman  DET   
778   1929      Johnny Evers                           Second Baseman  BSN   
797   1914      Frank Chance  First Baseman, Catcher and Rightfielder  NYY   
971   1945        Paul Waner                             Rightfielder  NYY   
1589  1922       Kiki Cuyler                               Outfielder  PIT   
3593  1979        Tim Raines                              Leftfielder  MON   

     games pa ab runs hits hr rbi sb cs bb so   ba  obp  slg  ops tb ibb  
85       1  0  0    0    0  0   0  0  0  0  0  NaN  NaN  NaN  NaN  0   0  
576      1  0  0    0    0  0   0  0  0  0  0  NaN  NaN  NaN  NaN  0   0  
580      1  0  0    0    0  0   0  0  0  0  0  NaN  NaN  NaN  NaN  0   0  
778      1  0  0    0    0  0   0  0  0  0  0  NaN  NaN  NaN  NaN  0   0  
797      1  0  0    0    0  0   0  0  0  0  0  NaN  NaN  NaN  NaN  0   0  
971      1  1  0    0    0  0   0  0  0  1  0  NaN    1  NaN  NaN  0   0  
1589     1  0  0    0    0  0   0  0  0  0  0  NaN  NaN  NaN  NaN  0   0  
3593     6  0  0    3    0  0   0  2  0  0  0  NaN  NaN  NaN  NaN  0   0

Remove missing BAs, very little data available in those observations

In [67]:
batters = batters.dropna(subset=['ba']) 

In [69]:
batters.loc[batters.isnull().slg,:]

year             name                                  position team  \
515   1899     Jim O'Rourke     Outfielder, Catcher and First Baseman  BPT   
567   1903  Hughie Jennings               Shortstop and First Baseman  BLT   
742   1885   Tommy McCarthy                                Outfielder  HVH   
945   1947      Jimmie Foxx  First Baseman, Third Baseman and Catcher  SPT   
1783  1911    Dave Bancroft                                 Shortstop  SUP   
2052  1879     Roger Connor           First Baseman and Third Baseman  HLY   
2905  1879       Ned Hanlon                             Centerfielder  ALB   

     games   pa   ab runs hits   hr  rbi   sb cs   bb   so     ba    obp  slg  \
515     92  355  355   50  102  NaN  NaN   14  0  NaN  NaN  0.287  0.287  NaN   
567     32  122  122   26   40  NaN  NaN    9  0  NaN  NaN  0.328  0.328  NaN   
742      7   28   28    7    7  NaN  NaN  NaN  0  NaN  NaN   0.25   0.25  NaN   
945      6    6    6  NaN    1  NaN  NaN  NaN  0  NaN  NaN  0.167  0.167  NaN   
1783   122  524  524  NaN  143  NaN  NaN  NaN  0  NaN  NaN  0.273  0.273  NaN   
2052    45  210  210   49   77  NaN  NaN  NaN  0  NaN  NaN  0.367  0.367  NaN   
2905    47  203  203   44   64  NaN  NaN  NaN  0  NaN  NaN  0.315  0.315  NaN   

      ops tb ibb  
515   NaN  0   0  
567   NaN  0   0  
742   NaN  0   0  
945   NaN  0   0  
1783  NaN  0   0  
2052  NaN  0   0  
2905  NaN  0   0

Removing missing slg's and ops's, only 7 rows and missing a bunch of other stats

In [70]:
batters = batters.dropna(subset=['slg']) 

In [74]:
batters.loc[batters.isnull().team,:]

year            name                        position team games  pa  ab  \
225   1910         Ty Cobb                   Centerfielder  NaN     5  23  20   
411   1912   Eddie Collins                  Second Baseman  NaN    11  46  36   
1178  1912  Home Run Baker                   Third Baseman  NaN    10  43  39   
1243  1910    Sam Crawford  Rightfielder and First Baseman  NaN    12  54  50   
1454  1908  Miller Huggins                  Second Baseman  NaN    12  53  43   

     runs hits hr rbi sb cs bb   so     ba    obp    slg    ops  tb ibb  
225     4    7  1   5  0  1  2    1   0.35  0.391    0.6  0.991  12   0  
411    14   15  0   6  5  0  7  NaN  0.417  0.512  0.528  1.039  19   0  
1178    4   15  0  11  2  0  2  NaN  0.385  0.395  0.641  1.036  25   0  
1243    7   18  0   7  0  1  3    1   0.36  0.396   0.46  0.856  23   0  
1454    7   13  0   3  2  0  8  NaN  0.302  0.412  0.302  0.714  13   0

In [75]:
batters = batters.dropna(subset=['team']) 

In [77]:
batters.loc[batters.isnull().any(axis=1),:]

year              name                                    position  \
18    1936     Willard Brown                                Rightfielder   
19    1937     Willard Brown                                Rightfielder   
20    1938     Willard Brown                                Rightfielder   
21    1939     Willard Brown                                Rightfielder   
22    1940     Willard Brown                                Rightfielder   
23    1941     Willard Brown                                Rightfielder   
24    1942     Willard Brown                                Rightfielder   
25    1943     Willard Brown                                Rightfielder   
26    1944     Willard Brown                                Rightfielder   
27    1946     Willard Brown                                Rightfielder   
28    1947     Willard Brown                                Rightfielder   
30    1948     Willard Brown                                Rightfielder   
31    1949     Willard Brown                                Rightfielder   
32    1950     Willard Brown                                Rightfielder   
84    1913  Billy Southworth                                Rightfielder   
87    1914  Billy Southworth                                Rightfielder   
88    1915  Billy Southworth                                Rightfielder   
91    1916  Billy Southworth                                Rightfielder   
92    1917  Billy Southworth                                Rightfielder   
93    1918  Billy Southworth                                Rightfielder   
107   1928  Billy Southworth                                Rightfielder   
108   1929  Billy Southworth                                Rightfielder   
110   1930  Billy Southworth                                Rightfielder   
111   1931  Billy Southworth                                Rightfielder   
112   1932  Billy Southworth                                Rightfielder   
113   1935  Billy Southworth                                Rightfielder   
174   1896      Honus Wagner   Shortstop, Rightfielder and First Baseman   
175   1897      Honus Wagner   Shortstop, Rightfielder and First Baseman   
219   1905           Ty Cobb                               Centerfielder   
249   1876     George Wright                Shortstop and Second Baseman   
...    ...               ...                                         ...   
2909  1883        Ned Hanlon                               Centerfielder   
2910  1884        Ned Hanlon                               Centerfielder   
2911  1885        Ned Hanlon                               Centerfielder   
2919  1945        Nellie Fox                              Second Baseman   
2921  1947        Nellie Fox                              Second Baseman   
2923  1948        Nellie Fox                              Second Baseman   
2942  1943        Larry Doby                               Centerfielder   
2943  1944        Larry Doby                               Centerfielder   
2944  1946        Larry Doby                               Centerfielder   
2945  1947        Larry Doby                               Centerfielder   
2958  1959        Larry Doby                               Centerfielder   
2982  1910      George Davis  Shortstop, Third Baseman and Centerfielder   
3003  1956    Orlando Cepeda               First Baseman and Leftfielder   
3078  1882        Bid McPhee                              Second Baseman   
3079  1883        Bid McPhee                              Second Baseman   
3080  1884        Bid McPhee                              Second Baseman   
3081  1885        Bid McPhee                              Second Baseman   
3082  1886        Bid McPhee                              Second Baseman   
3320  1937        Joe Gordon                              Second Baseman   
3429  1876      Deacon White     Third Baseman, Catcher and Rightfielder   
3430  1877      Deacon White     Third Baseman, Catche

Before converting, fill NAs with a invalid number (I can't think of how else to do this besides disregarding missing data). Imputing random values seems inaccurate, and most of the counting stats are random and independent from their rates. I'll fill them in with an invalid number (-999), that way I know that they were missing. 

In [85]:
batters['bb'] = batters['bb'].fillna(-999)

In [91]:
batters['bb'] = batters['bb'].astype(int)
batters['pa'] = batters['pa'].astype(int)
batters['bb_rate'] = batters['bb'] / batters['pa']

In [95]:
# Now so rate
batters['so'] = batters['so'].fillna(-999)

In [97]:
batters['so'] = batters['so'].astype(int)

In [98]:
batters['so_rate'] = batters['so'] / batters['pa']

In [100]:
batters

year           name       position     team games   pa   ab runs hits  \
0     1990   Jeff Bagwell  First Baseman  BOS-min   136  569  481   63  160   
1     1991   Jeff Bagwell  First Baseman      HOU   156  650  554   79  163   
2     1992   Jeff Bagwell  First Baseman      HOU   162  697  586   87  160   
3     1993   Jeff Bagwell  First Baseman      HOU   142  609  535   76  171   
4     1994   Jeff Bagwell  First Baseman      HOU   110  479  400  104  147   
5     1995   Jeff Bagwell  First Baseman  HOU-min     4   16   12    0    2   
6     1995   Jeff Bagwell  First Baseman      HOU   114  539  448   88  130   
7     1996   Jeff Bagwell  First Baseman      HOU   162  719  568  111  179   
8     1997   Jeff Bagwell  First Baseman      HOU   162  717  566  109  162   
9     1998   Jeff Bagwell  First Baseman      HOU   147  661  540  124  164   
10    1999   Jeff Bagwell  First Baseman      HOU   162  729  562  143  171   
11    2000   Jeff Bagwell  First Baseman      HOU   159  719  590  152  183   
12    2001   Jeff Bagwell  First Baseman      HOU   161  717  600  126  173   
13    2002   Jeff Bagwell  First Baseman      HOU   158  691  571   94  166   
14    2003   Jeff Bagwell  First Baseman      HOU   160  702  605  109  168   
15    2004   Jeff Bagwell  First Baseman      HOU   156  679  572  104  152   
16    2005   Jeff Bagwell  First Baseman  HOU-min     3   12    9    1    2   
17    2005   Jeff Bagwell  First Baseman      HOU    39  123  100   11   25   
18    1936  Willard Brown   Rightfielder      KCM   NaN    1    1    0    0   
19    1937  Willard Brown   Rightfielder      KCM   NaN  178  174   36   58   
20    1938  Willard Brown   Rightfielder      KCM   NaN   78   78   14   27   
21    1939  Willard Brown   Rightfielder      KCM   NaN   72   70   10   22   
22    1940  Willard Brown   Rightfielder      KCM   NaN    4    4    0    0   
23    1941  Willard Brown   Rightfielder      KCM   NaN   92   87   19   30   
24    1942  Willard Brown   Rightfielder      KCM   NaN  133  127   28   47   
25    1943  Willard Brown   Rightfielder      KCM   NaN   79   79   15   27   
26    1944  Willard Brown   Rightfielder      KCM   NaN    4    4    0    1   
27    1946  Willard Brown   Rightfielder      KCM   NaN  109  106   15   27   
28    1947  Willard Brown   Rightfielder      KCM   NaN   61   57   12   21   
29    1947  Willard Brown   Rightfielder      SLB    21   67   67    4   12   
...    ...            ...            ...      ...   ...  ...  ...  ...  ...   
3594  1980     Tim Raines    Leftfielder  MON-min   108  497  429  105  152   
3595  1980     Tim Raines    Leftfielder      MON    15   27   20    5    1   
3596  1981     Tim Raines    Leftfielder      MON    88  363  313   61   95   
3597  1982     Tim Raines    Leftfielder      MON   156  731  647   90  179   
3598  1983     Tim Raines    Leftfielder      MON   156  720  615  133  183   
3599  1984     Tim Raines    Leftfielder      MON   160  718  622  106  192   
3600  1985     Tim Raines    Leftfielder      MON   150  665  575  115  184   
3601  1986     Tim Raines    Leftfielder      MON   151  664  580   91  194   
3602  1987     Tim Raines    Leftfielder      MON   139  627  530  123  175   
3603  1988     Tim Raines    Leftfielder      MON   109  488  429   66  116   
3604  1989     Tim Raines    Leftfielder      MON   145  618  517   76  148   
3605  1990     Tim Raines    Leftfielder      MON   130  538  457   65  131   
3606  1991     Tim Raines    Leftfielder      CHW   155  709  609  102  163   
3607  1992     Tim Raines    Leftfielder      CHW   144  644  551  102  162   
3608  1993     Tim Raines    Leftfielder  CHW-min     3   13   11    3    5   
3609  1993     Tim Raines    Leftfielder      CHW   115  486  415   75  127   
3610  1994     Tim Raines    Leftfielder      CHW   101  453  384   80  102   
3611  1995     Tim Raines    Leftfielder      CHW   133  581  502   81  143   
3612  1996     Tim Raines    Leftfielder  NYY-min    22 

Write this clean dataframe before everything crashes and dies!!!

In [103]:
batters.to_csv('batters_final.csv', header = True)